inox2d/src/render.rs

```diff
@@ -218,3 +222,173 @@ impl Puppet {
+                    ) => {
+                        self.draw_composite(true, camera, mask_composite, puppet, mask_children)?;
+                    }
+
+                    _ => {
+                        // This match block clearly is sign that the data structure needs rework
+                        todo!();
+                    }
+                }
+            }
+            self.on_begin_masked_content()?;
+            self.draw_part_self(false, camera, node_render_ctx, part, part_render_ctx)?;
+            self.on_end_mask()?;
+        } else {
+            self.draw_part_self(false, camera, node_render_ctx, part, part_render_ctx)?;
+        }
+        Ok(())
+    }
+
+    fn begin_composite(&self) -> Result<(), Self::Error>;
+    fn end_composite(&self) -> Result<(), Self::Error>;
+
+    fn draw_composite_self(
+        &self,
+        as_mask: bool,
+        camera: &Mat4,
+        puppet: &Puppet,
+        children: &[InoxNodeUuid],
+    ) -> Result<(), Self::Error> {
+        self.begin_composite()?;
+
+        for &uuid in children {
+            let node = puppet.nodes.get_node(uuid).unwrap();
+            let node_render_ctx = &puppet.render_ctx.node_render_ctxs[&uuid];
+
+            if let (InoxData::Part(ref part), RenderCtxKind::Part(ref part_render_ctx)) =
+                (&node.data, &node_render_ctx.kind)
+            {
+                if as_mask {
+                    self.draw_part_self(true, camera, node_render_ctx, part, part_render_ctx)?;
+                } else {
+                    self.draw_part(camera, node_render_ctx, part, part_render_ctx, puppet)?;
+                }
+            } else {
+                // composite inside composite simply cannot happen
+            }
+        }
+
+        self.end_composite()?;
+        Ok(())
+    }
+
+    fn draw_composite(
+        &self,
+        as_mask: bool,
+        camera: &Mat4,
+        composite: &Composite,
+        puppet: &Puppet,
+        children: &[InoxNodeUuid],
+    ) -> Result<(), Self::Error>;
+
+    fn draw_drawables(&self, camera: &Mat4, puppet: &Puppet) -> Result<(), Self::Error> {
+        for &uuid in &puppet.render_ctx.drawables_zsorted {
+            let node = puppet.nodes.get_node(uuid).unwrap();
+            let node_render_ctx = &puppet.render_ctx.node_render_ctxs[&uuid];
+
+            match (&node.data, &node_render_ctx.kind) {
+                (InoxData::Part(ref part), RenderCtxKind::Part(ref part_render_ctx)) => {
+                    self.draw_part(camera, node_render_ctx, part, part_render_ctx, puppet)?;
+                }
+
+                (InoxData::Composite(ref composite), RenderCtxKind::Composite(ref children)) => {
+                    self.draw_composite(false, camera, composite, puppet, children)?;
+                }
+
+                _ => {
+                    // This clearly is sign that the data structure needs rework
+                    todo!();
+                }
+            }
```

Comment:
I think we can manage without for now. Maybe this match statement could be its own method from the `InoxRenderer` trait with this as the default implementation, something like `fn draw_node(&self, node: &InoxNode, node_render_ctx: &RenderCtx, camera: &Mat4, puppet: &Puppet)`?

Though the more I look at this, the more I feel like a way to make this flexible for any kind of node variant or custom type would be to turn nodes into a trait again and abstract the drawing process there... But I don't know if it's a good idea.

Reply:
I suggest rework because `PartRenderCtx` info is always coupled with `Part` and `CompositeRenderCtx` is always coupled with `Composite`. Because of we insisted on separating them while they are actually coupled, we are now left with the `_ => ()` branch, signaling two meaningless combinations.

I am suggesting putting `RenderCtx` info into `Part` and `Composite`. They are now separated because in the renderless API refactor by me I factored out all info directly used for rendering. But we are inox2d and every data structure is directly or indirectly related to the render of a puppet, so really no point to make the distinction. And indexing by `uuid` twice to retrieve two pieces of info takes twice the overhead.

By the way, please note that in Rust `enum`s are implemented as tagged unions, so in future refactors please mind not to make different variants of an enum taking too different spaces or we will be wasting a lot space on padding.

Reply:
Oh I forgot to reply to the `draw_node` part: The thought of factoring out this `match` appeared again and again through out the development process, however the two places that deploy this match have slight differences which make such a function useless (you will need two `bool`s for this function). The conclusion I reached is that it is because the currently segmented data structure, a simple access to drawable data from one node is tedious so that one will try to factor this process out. But if the data structure is reworked maybe this will no longer be the case.

Reply:
> But we are inox2d and every data structure is directly or indirectly related to the render of a puppet, so really no point to make the distinction.

I guess we can trade off some amount of memory in nodes and gain back some inside the renderer since we won't need the hashmap anymore. You're not wrong, trying to separate the two has negligible benefits given the use-case of inox2d.



> By the way, please note that in Rust `enum`s are implemented as tagged unions, so in future refactors please mind not to make different variants of an enum taking too different spaces or we will be wasting a lot space on padding.

Don't worry, I always keep that in mind. :p